lib/zzhc/zzhc_port.c

```diff
@@ -0,0 +1,365 @@
+{
+	settings_load();
+	memcpy(iccid_buf, ueiccid, buf_len);
+	return 0;
+}
+
+/**
+ * @brief Function to send an AT command and receive response immediately
+ *
+ * This function should be used if the response from the modem should be
+ * returned in a user supplied buffer. This function will return an empty buffer
+ * if nothing is returned by the modem.
+ *
+ * @param cmd Pointer to null terminated AT command string
+ * @param buf Buffer to put the response in. NULL pointer is allowed, see
+ *            behaviour explanation for @ref buf_len equals 0.
+ * @param buf_len Length of response buffer. 0 length is allowed and will send
+ *                the command, process the return code from the modem, but
+ *                any returned data will be dropped.
+ *
+ * @retval 0 If command execution was successful (same as OK returned from
+ *           modem). ERROR will return ENOEXEC (positve).
+ *
+ * @retval -ENOBUFS is returned if AT_CMD_RESPONSE_MAX_LEN is not large enough
+ *         to hold the data returned from the modem.
+ * @retval ENOEXEC is returned if the modem returned ERROR.
+ * @retval -EMSGSIZE is returned if the supplied buffer is to small or NULL.
+ * @retval -EIO is returned if the function failed to send the command.
+ *
+ */
+static int at_cmd_xfer(const char *const cmd, char *buf, size_t buf_len)
+{
+	return at_cmd_write(cmd, buf, buf_len, NULL);
+}
+
+/**@brief Register AT-notification handler */
+static void register_handler(void *context, void *handler)
+{
+	at_notif_register_handler(context, handler);
+}
+
+/**@brief De-register AT-notification handler */
+static void deregister_handler(void *context, void *handler)
+{
+	at_notif_deregister_handler(context, handler);
+}
+
+/**@brief Encode a buffer into base64 format.
+ *
+ * This function is equivalent to bash command "base64 < src.bin > dst.bin"
+ *
+ * @param dst      destination buffer
+ * @param dlen     size of the destination buffer
+ * @param src      source buffer
+ * @param slen     amount of data to be encoded
+ *
+ * @return         Number of bytes encoded.
+ */
+static int base64(unsigned char *dst, size_t dlen, const unsigned char *src,
+		size_t slen)
+{
+	int i = 0, rc;
+
+	if (dlen == 0 || slen == 0) {
+		return 0;
+	}
+
+	rc = base64_encode(dst, dlen, &i, src, slen);
+	return (rc == 0) ? i : 0;
+}
+
+
+static int settings_set(const char *key, size_t len,
+		settings_read_cb read_cb, void *cb_arg)
+{
+	int rc;
+
+	if (strcmp(key, "iccid") == 0) {
+		rc = read_cb(cb_arg, ueiccid, ICCID_LEN);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+/**@brief Initialization of external modules */
+static int ext_init(struct zzhc *ctx)
+{
+	int rc;
+	struct settings_handler *this_settings;
+
+	memset(ueiccid, 0, ICCID_LEN);
+
+	/* Allocate memory for thread, sem, at_list and storage */
+	ctx->thread = this_malloc(sizeof(struct k_thread));
+	if (ctx->thread == NULL) {
+		return -ENOBUFS;
+	}
+
+	ctx->sem = this_malloc(sizeof(struct k_sem));
```

Comment:
Do w really need to use heap for all these allocations?

Reply:
Perhaps we can continue discussion [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339536848).